packages/gensx-openai/src/composition.tsx

```diff
@@ -0,0 +1,359 @@
+});
+
+// Stream transform component
+type StreamTransformProps = RawCompletionProps & {
+  stream: true;
+  tools?: GSXTool<z.ZodObject<z.ZodRawShape>>[];
+};
+
+type StreamTransformReturn = Stream<ChatCompletionChunk>;
+
+export const StreamTransform = gsx.Component<
+  StreamTransformProps,
+  StreamTransformReturn
+>("StreamTransform", async (props) => {
+  const context = gsx.useContext(OpenAIContext);
+  if (!context.client) {
+    throw new Error(
+      "OpenAI client not found in context. Please wrap your component with OpenAIProvider.",
+    );
+  }
+
+  const { stream, tools, ...rest } = props;
+  const openAITools = tools?.map((tool) => tool.toOpenAITool());
+
+  return context.client.chat.completions.create({
+    ...rest,
+    tools: openAITools,
+    stream: true,
+  } as ChatCompletionCreateParamsStreaming);
+});
+
+// Tool execution component
+interface ToolExecutorProps {
+  tools: GSXTool<z.ZodObject<z.ZodRawShape>>[];
+  toolCalls: NonNullable<
+    ChatCompletionOutput["choices"][0]["message"]["tool_calls"]
+  >;
+  messages: ChatCompletionMessageParam[];
+}
+
+type ToolExecutorReturn = ChatCompletionOutput;
+
+export const ToolExecutor = gsx.Component<
+  ToolExecutorProps,
+  ToolExecutorReturn
+>("ToolExecutor", async (props) => {
+  const { tools, toolCalls, messages } = props;
+  const context = gsx.useContext(OpenAIContext);
+  if (!context.client) {
+    throw new Error(
+      "OpenAI client not found in context. Please wrap your component with OpenAIProvider.",
+    );
+  }
+
+  // Execute each tool call
+  const toolResults = await Promise.all(
+    toolCalls.map(async (toolCall) => {
+      const tool = tools.find((t) => t.name === toolCall.function.name);
+      if (!tool) {
+        throw new Error(`Tool ${toolCall.function.name} not found`);
+      }
+
+      try {
+        const args = JSON.parse(toolCall.function.arguments) as z.infer<
+          typeof tool.parameters
+        >;
+        const result = await tool.execute(args);
+        return {
+          tool_call_id: toolCall.id,
+          role: "tool" as const,
+          content: typeof result === "string" ? result : JSON.stringify(result),
+        };
+      } catch (e) {
+        throw new Error(
+          `Failed to execute tool ${toolCall.function.name}: ${
+            e instanceof Error ? e.message : String(e)
+          }`,
+        );
+      }
+    }),
+  );
+
+  // Add tool results to messages and make another completion
+  const updatedMessages = [...messages, ...toolResults];
+
+  const completion = await context.client.chat.completions.create({
+    messages: updatedMessages,
+    stream: false,
+  } as ChatCompletionCreateParamsNonStreaming);
+
+  return completion;
+});
+
+// Tool transform component
+type ToolTransformProps = RawCompletionProps & {
+  tools: GSXTool<z.ZodObject<z.ZodRawShape>>[];
+};
+
+type ToolTransformReturn = RawCompletionReturn;
+
+export const ToolTransform = gsx.Component<
```

Comment:
I think this transform naming is throwing me off. Doesn't this actually run the tools if you use it?